### DISCUSSION TITLE
A parallelization problem: We are working on a project to train robots in webots through reinforcement learning

### DISCUSSION THREAD

**Discussion by zyegao** [cyberbotics/webots#6519] [Q&A]
We have a reinforcement learning project where we created our own robot in webots and trained it to complete tasks through reinforcement learning. But now we encounter a problem that cannot be solved, because the reinforcement learning task requires the robot to complete a lot of data collection. However, currently we can only start a webots environment in a Linux machine, and then use python code to control the robot in the environment. In other words, there is only one environment that can be used to collect data in the reinforcement learning task, so the entire algorithm will take a very long time.

So we want to ask for help. Is it possible to open multiple webots environments on the same machine, and then control the robots in each environment separately through multiple processes in one python code. Very much looking forward to getting your answers

**Comment by omichel** [comment]
Yes, it is possible to launch several instances of Webots in parallel on a Linux machine and have extern controllers connect to a specific instance.

**Reply by zyegao** [reply]
But how to achieve it. Is it by matching the IP and port number?

**Reply by omichel** [reply]
Yes, exactly.

**Comment by FrostyTheSouthernSnowman** [comment]
You could also instantiate multiple instances of your robot and environment to train in parallel, like if you have a room it’s training in, create a grid of separate copies of that room and place an agent in each one so they all run in parallel from one website instance, similar to how people often do it in software like unity. See [this video](https://youtu.be/zPFU30tbyKs?si=1SYSQCa6GyQ3NLyq) around the 33:45 timestamp.

**Comment by yeko31** [comment]
Hi
May I ask whether is it possible to deploy the trained reinforcement learning algorithm in webots(using deepbots) on a real hardware platform to test it in real environmental conditions? for a drone obstacle avoidance problem?
Appreciate a comment please
